src/transformers/modeling_utils.py

```diff
@@ -147,6 +148,47 @@ def find_tensor_attributes(module: nn.Module) -> List[Tuple[str, Tensor]]:
         return first_tuple[1].dtype
 
 
+def get_parameter_first_float_dtype_or_any_dtype(parameter: Union[nn.Module, GenerationMixin, "ModuleUtilsMixin"]):
+    """returns the first found floating dtype in parameters if there is one, otherwise returns the first dtype it finds"""
+    try:
+        for t in parameter.parameters():
+            if t.is_floating_point():
+                return t.dtype
+
+    except StopIteration:
+        # For nn.DataParallel compatibility in PyTorch 1.5
+
+        def find_tensor_attributes(module: nn.Module) -> List[Tuple[str, Tensor]]:
+            tuples = [(k, v) for k, v in module.__dict__.items() if torch.is_tensor(v)]
+            return tuples
+
+        gen = parameter._named_members(get_members_fn=find_tensor_attributes)
+        for tuple in gen:
+            if tuple[1].is_floating_point():
+                return tuple[1].dtype
+        # fallback to any dtype the model has even if not floating
+        return tuple[1].dtype
+
+
+def get_state_dict_first_float_dtype(state_dict):
+    """returns the first found floating dtype in state_dict or asserts if none were found"""
+    for t in state_dict.values():
+        if t.is_floating_point():
+            return t.dtype
+
+    raise ValueError("couldn't find any floating point dtypes in state_dict")
+
+
+def get_state_dict_first_float_dtype_or_any_dtype(state_dict):
```

Comment:
this is a super ambiguous name since it tells very little to the user of what it does and suggests that the whole state_dict has a single dtype, which was our initial wrong assumption.

But I'm happy to oblige if the more explicit name is too long.

Reply:
It comes under the assumption that *most* of the state dict is of the same type except some outliers. The case we had doesn't have half its parameters in the int dtype, just one buffer.